### PR TITLE
Adds MFA Duo configuration

### PR DESCRIPTION
**Pull request by petems**
Adds ability to configure MFA with Duo: https://www.vaultproject.io/docs/enterprise/mfa/mfa-duo.html

### CODE REVIEW

File: vault/resource_mfa_duo_test.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+	"os"
+
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/hashicorp/vault/api"
+)
+
+func testMFADuo_basic(t *testing.T) {
```
Comment (tyrannosaurus-becks) [inline vault/resource_mfa_duo_test.go]:
```suggestion
func TestMFADuo_basic(t *testing.T) {
```

Comment (tyrannosaurus-becks) [reply vault/resource_mfa_duo_test.go]: I found that when I run this test, I get:
```
=== RUN   TestMFADuo_basic
--- FAIL: TestMFADuo_basic (0.01s)
    testing.go:538: Step 0 error: config is invalid: vault_mfa_duo.test: "push_info": required field is not set
FAIL

Process finished with exit code 1
```

File: vault/resource_mfa_duo.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+				Required:     true,
+				Description:  "Name of the MFA method.",
+				ValidateFunc: validateNoTrailingSlash,
+			},
+			"mount_accessor": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The mount to tie this method to for use in automatic mappings. The mapping will use the Name field of Aliases associated with this mount as the username in the mapping.",
+			},
+			"username_format": {
```
Comment (tyrannosaurus-becks) [inline vault/resource_mfa_duo.go]: It would be great to get test coverage of this field, as well as `push_info`.

Comment (petems) [reply vault/resource_mfa_duo.go]: Done! 👍

File: vault/resource_mfa_duo_test.go (hunk 1 above)
Comment (petems) [reply vault/resource_mfa_duo_test.go]: Ok, I think I've fixed it 😄

**Comment by petems** [comment]
Hmm, test failure is for `TestDataSourcePolicyDocument`... can we run again?

**Comment by tyrannosaurus-becks** [comment]
Looks like I get a test failure when I run them locally against Vault enterprise!
```
$ go test -v -run TestMFADuoBasic
=== RUN   TestMFADuoBasic
--- FAIL: TestMFADuoBasic (0.06s)
    testing.go:568: Step 0 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        UPDATE: vault_mfa_duo.test
          api_hostname:    "api-2b5c39f5.duosecurity.com" => "api-2b5c39f5.duosecurity.com"
          id:              "mfa-duo-4083353440263821461" => "mfa-duo-4083353440263821461"
          integration_key: "" => "BIACEUEAXI20BNWTEYXT"
          mount_accessor:  "auth_userpass_59c73587" => "auth_userpass_59c73587"
          name:            "mfa-duo-4083353440263821461" => "mfa-duo-4083353440263821461"
          push_info:       "" => "from=loginortal&domain=example.com"
          secret_key:      "" => "8C7THtrIigh2rPZQMbguugt8IUftWhMRCOBzbuyz"
          username_format: "user@example.com" => "user@example.com"
        
        
        
        STATE:
        
        vault_auth_backend.userpass:
          ID = userpass-6631647075597508582
          provider = provider.vault
          accessor = auth_userpass_59c73587
          default_lease_ttl_seconds = 0
          description = 
          listing_visibility = 
          local = false
          max_lease_ttl_seconds = 0
          path = userpass-6631647075597508582
          type = userpass
        vault_mfa_duo.test:
          ID = mfa-duo-4083353440263821461
          provider = provider.vault
          api_hostname = api-2b5c39f5.duosecurity.com
          integration_key = 
          mount_accessor = auth_userpass_59c73587
          name = mfa-duo-4083353440263821461
          push_info = 
          secret_key = 
          username_format = user@example.com
        
          Dependencies:
            vault_auth_backend.userpass
    testing.go:629: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: config is invalid: cannot write to a path ending in '/'
        
        State: vault_auth_backend.userpass:
          ID = userpass-6631647075597508582
          provider = provider.vault
          accessor = auth_userpass_59c73587
          default_lease_ttl_seconds = 0
          description = 
          listing_visibility = 
          local = false
          max_lease_ttl_seconds = 0
          path = userpass-6631647075597508582
          type = userpass
        vault_mfa_duo.test:
          ID = mfa-duo-4083353440263821461
          provider = provider.vault
          api_hostname = api-2b5c39f5.duosecurity.com
          integration_key = 
          mount_accessor = auth_userpass_59c73587
          name = mfa-duo-4083353440263821461
          push_info = 
          secret_key = 
          username_format = user@example.com
        
          Dependencies:
            vault_auth_backend.userpass
FAIL
exit status 1
FAIL	github.com/terraform-providers/terraform-provider-vault/vault	0.069s
```

File: vault/resource_mfa_duo_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+resource "vault_auth_backend" "userpass" {
+  type = "userpass"
+  path = %q
+}
+
+resource "vault_mfa_duo" "test" {
+  name                  = %q
+  mount_accessor        = "${vault_auth_backend.userpass.accessor}"
+  secret_key            = "8C7THtrIigh2rPZQMbguugt8IUftWhMRCOBzbuyz"
+  integration_key       = "BIACEUEAXI20BNWTEYXT"
```
Comment (tyrannosaurus-becks) [inline vault/resource_mfa_duo_test.go]: I assume these are fake but just wanted to point them out in case.

Comment (petems) [reply vault/resource_mfa_duo_test.go]: Yep, I took these straight from the API docs: https://www.vaultproject.io/api/system/mfa/duo.html#sample-payload

Comment (tyrannosaurus-becks) [reply vault/resource_mfa_duo_test.go]: :+1: that's what I thought! Thanks for confirming!

File: vault/resource_mfa_duo.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+			},
+			"secret_key": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Secret key for Duo.",
+			},
+			"integration_key": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Integration key for Duo.",
```
Comment (tyrannosaurus-becks) [inline vault/resource_mfa_duo.go]: Also, I just realized the secret key and integration key might need to be marked "Sensitive" so they don't pop out in Terraform's output.

Comment (petems) [reply vault/resource_mfa_duo.go]: Done! 👍

**Comment by petems** [comment]
@tyrannosaurus-becks The tests should pass now! The API wont return the secret key and integration key, so it was overwriting them as blank second time round.

**Pull request merged**